src/ImageSharp/Advanced/AdvancedImageExtensions.cs

```diff
@@ -24,78 +24,134 @@ public static Configuration GetConfiguration<TPixel>(this Image<TPixel> source)
             => GetConfiguration((IConfigurable)source);
 
         /// <summary>
-        /// Returns a reference to the 0th element of the Pixel buffer,
-        /// allowing direct manipulation of pixel data through unsafe operations.
-        /// The pixel buffer is a contiguous memory area containing Width*Height TPixel elements laid out in row-major order.
+        /// Gets the representation of the pixels as a <see cref="Memory{T}"/> of contiguous memory in the source image's pixel format
+        /// stored in row major order.
         /// </summary>
         /// <typeparam name="TPixel">The Pixel format.</typeparam>
-        /// <param name="source">The source image frame</param>
-        /// <returns>A pinnable reference the first root of the pixel buffer.</returns>
-        public static ref TPixel DangerousGetPinnableReferenceToPixelBuffer<TPixel>(this ImageFrame<TPixel> source)
+        /// <param name="source">The source <see cref="ImageFrame{TPixel}"/></param>
+        /// <returns>The <see cref="Memory{T}"/></returns>
+        public static Memory<TPixel> GetPixelMemory<TPixel>(this ImageFrame<TPixel> source)
             where TPixel : struct, IPixel<TPixel>
-         => ref DangerousGetPinnableReferenceToPixelBuffer((IPixelSource<TPixel>)source);
+        {
+            return source.PixelBuffer.Buffer.Memory;
```

Comment:
we shouldn't expose the buffer as a Memory<T> as we then can't swap the buffer and return the memory to the Array pool other wise someone could have a reference to a piece of memory that has been returned to someone else or was owned by an external process and it could leak dangerous things

Reply:
The image itself should be treated as Memory<T> of sorts but can never share the hand over directly indefinitely.

Reply:
@tocsoft the same could be true for `Span<T>`:

```C#
Span<int> testSpan = default;

using (IBuffer<T> buffer = allocator.Allocate<int>(10))
{
    // span is assigned, we are on the same method stack in the using block:
    testSpan = buffer.GetSpan(); 
}

using (IBuffer<T> somethingElse= allocator.Allocate<int>(10))
{
     somethingElse.GetSpan()[5] = 666;
}

// we are out of the intended scope here:
int naiveUsersValue = testSpan[5];
```

Despite the dangers I think both `image.GetPixelMemory()` and `image.GetPixelRowMemory(y)` are useful methods for advanced users, because they allow heaping pixel rows, so users can use them in async API-s etc.

What we can do is:
1. Document these pitfals
2. Introduce a new level of indirection: Decouple `IBuffer<T>` from `MemoryManager<T>`, so we can swap the buffer behind the `Memory<T>` instance.

Implementing point 2. would take many hours, delaying the open-up of memory API-s further. We can actually do it later, as an additional increment.

Reply:
We have no one actively after Memory<T> when Span<T> as having a reference to the image is there. I vote we remove it from the API until we can expose it safely or at least its requested. Until then its an API we have to support which very easily (unlike span which is much harder and can only leak within a single stack) can leak the contents of memory at any random point in the future of the whole application.

I think its too much risk with near to zero benefits that can be handled by keeping a reference to the Image.

Reply:
@tocsoft you can't do the following with span:
```C#
async Task DoOperationOnPixelsAsync(Span<Rgba32> pixels)
{
    await .....;
    await ........;
}
```

One of our users (on gitter) walked into this pitfall resulting in a runtime failure, and he blamed ImageSharp for it. (This might be fine, we can't expect everyone to understand stack-only types.) We decided to hide the `Span<T>` API-s because of this story. There was a quite valid reasoning, that we should expose only the `Memory<T>` variant.

I say we should expose either both the Span and the Memory API-s or neither of them. If we can't agree, I say let's internalize everything for now, turning this PR into a refactor one.

@JimBobSquarePants your thoughts?

Reply:
Actually ... you don't even need to have a goal of implementing something in a separate async operation for let's say perf. If you are implementing an ASP.NET controller, being in an async context most of the time, it's very easy to walk into a pitfall of this kind.

Reply:
1.  `Span<T>` is now hidden away in the advanced namespace so we are much less likely to get people confused about using Span with async etc. 
2. you won't be doing pixel level operations over spans in a controller in MVC.. they will be advised to create a Processor and do there stuff there (MVC sorted)
3. if you changes `Span<T>`  in `DoOperationOnPixelsAsync` with `Image<T>` you can now work async and only have a manor (aggressively in lines called over to get the `Span<T>` as required)

If you are using the API's that expose `Span<T>` we are deeming the user knowledgeable enough about the framework to handle/understand how async and Spans don't mix well.

If we don't expose `Span<T>` then users can't write new `IImageProcessor`s they can do that without `Memory<T>` as that's exactly what we are doing internally.

Reply:
@mellinoe sorry for pulling you in, but I think your opinion might be very useful here!
@iamcarbon - same for you :)

TLDR:
- Exposing only `Span<TPixel>` API-s is dangerous, because it may lead to heap captures for users who don't follow (or don't know about) coding patterns we suggest.
- Exposing `Memory<TPixel>` API-s is dangerous, because the memory may point to an "outdated" location after certain processors being executed.

Reply:
@tocsoft if no one disagrees, I'm being tempted to accept your reasoning.

Exposing `Memory<T>` could be an incremental addition later.